[DI] Add a check on the getReflectionClass call

Taluu opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | ~
| License       | MIT
| Doc PR        | ~

This prevents from doing a getter injection if the definition's idcannot be resolved to an existing class.

Taluu committed:
[DI] Add a check on the getReflectionClass call

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php:
```diff
@@ -502,6 +502,11 @@ private function addServiceMethodCalls($id, Definition $definition, $variableNam
         $class = $this->container->getReflectionClass($definition->getClass());
+
+        if (!$class) {
+            throw new RuntimeException(sprintf('Unable to configure service "%s": class could not be reflected.', $id));
```
`[...] class "%s" not found.`?
can you review the other uses of getReflectionClass in this class and also in ContainerBuiler please?

Taluu replied:
Hum, I was thinking that maybe "could not be reflected" would be better, if there are other reasons than "not finding the class". But I can change it.

I'll check where the other calls are made.

nicolas-grekas replied:
`class "%s" not found.` is the only reason why it can fail - and "be reflected" will be obscure to people I guess, so I'm still up for the change

Taluu commented:
@nicolas-grekas I added a few checks here and there, but it always was in `PhpDumper`. the other ones (such as `ContainerBuilder`) do not really depend on the this method (Only `ContainerBuilder` seems to call this method, but with no implication on what it should return).

I added some basic tests (for when trying to dump a value), but I'm not sure how to test some other cases (such as adding a new instance or dumping a value...), as it's a lot of privates...

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php:
```diff
@@ -1645,7 +1656,11 @@ private function dumpValue($value, $interpolate = true)
                 throw new InvalidArgumentException(sprintf('Cannot create closure-proxy for service "%s": method "%s::%s" does not exist.', $reference, $class, $method));
             }
-            $r = $this->container->getReflectionClass($class)->getMethod($method);
+            $r = $this->container->getReflectionClass($class);
```
this one has already been checked with the method_exists check just above

Taluu replied:
Yes, but once again, I'm not sure if the reflection could be made for a reason X or Y. Otherwise, There is indeed only one place to change (what was done in the first version of this PR)

nicolas-grekas replied:
there is no other reason: when a class is already loaded, reflection can't fail

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php:
```diff
@@ -864,7 +869,13 @@ private function addNewInstance(Definition $definition, $return, $instantiation,
             $class = 'SymfonyProxy_'.md5($inheritanceProxy);
             $this->inheritanceProxies[$class] = $inheritanceProxy;
-            $constructor = $this->container->getReflectionClass($definition->getClass())->getConstructor();
+            $r = $this->container->getReflectionClass($definition->getClass());
```
this one is after the call to addServiceOverriddenMethods so it can also be removed

Taluu commented:
Changes made. So basically only where I triggered the error seemed to matter, as the rest was already checking the class existence (and thus could not fail the `getReflectionClass`. I modified the error message too.

nicolas-grekas reviewed (approved):
:+1:

xabbuh commented on src/Symfony/Component/DependencyInjection/Tests/Dumper/PhpDumperTest.php:
```diff
@@ -408,6 +408,23 @@ public function provideBadOverridenGetters()
+     * @expectedException \RuntimeException
+     * @expectedExceptionMessage Unable to configure service "Acme\FooNonExistant": class "Acme\FooNonExistant" not found.
+     */
+    public function testDumpOverriddenGetterOnNonExistantClassTriggersException()
```
Existent

fabpot replied:
d236af6757

xabbuh reviewed (approved):
👍 (existant could be replaced consistently with existent though)

fabpot commented:
Thank you @Taluu.
